EIPS/eip-3102.md

```diff
@@ -0,0 +1,189 @@
+
+#### Accessing account's balance, nonce, code, storage root and storage slots
+
+Assuming an account $A \equiv \left(A_b, A_n, A_c, A_s\right)$ at address $A_a$, the following elements can be found at the following addresses:
+
+ * The account balance $A_b \in \mathbb{B}_{32}$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 00_2$;
+ * The account nonce $A_n \in \mathbb{B}_8$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 01_2$;
+ * The code $A_c \in \mathbb{B}$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 10_2$;
+ * The root of the storage trie $A_s$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 11_2$
+     * The storage slot number $k \in \mathbb{B}_{32}$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 11_2 \space\texttt{++}\space \texttt{H}(k)$;
+
+After [EIP-2926](https://eips.ethereum.org/EIPS/eip-2926) has been rolled out, $A_c$ will represent the root of the code merkelization tree. The key accessing code chunk $c$ is $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 10_2 \space\texttt{++}\space \texttt{U256}(c)$.
+
+### Node merkelization rule
+
+Leaves and nodes that have no prefix are hashed according to the rule below:
+
+```
+internal_hash = hash(left_child_hash || right_child_hash)
+leaf_hash = hash(hash(key) || hash(leaf_value))
+```
+
+If a prefix is present, the length of the path from the root to the prefixed node is further concatenated with the output of the prefix-less rule, and hashed again:
+
+```
+internal_hash_with_prefix = hash(u256(path_length_u256 - 1) || internal_hash)
+leaf_hash_with_prefix = hash(u256(path_length_u256 - 1) || leaf_hash)
+```
+
+## Rationale
+
+### BLAKE2-256
+
+BLAKE2 offers better performance, which is key to compensate for the loss of performance associated to a ~4x increase in the number of nodes.
+
+BLAKE3 offers even better performance. No official golang release exists at the time of the writing of this document. This presents a security risk, and therefore BLAKE2 is considered instead.
+
+### Merging of the account and storage tries
+
+The trend in clients is to store the keys and values in a "flat" database. Having the key of any storage slot prefixed with the address key of the account it belongs to helps grouping all of an account's data on disk, as well as simplifying the witness structure.
+
+### Prefixes and extension nodes
+
+An alternative proposal has been made, which provides optimal witnesses. The trade-off is that extension nodes must be removed.
+
+```
+node_hash = hash(left_child_hash || right_child_hash)
+leaf_hash = hash(0 || leaf_value)
+```
+
+The removal of extension nodes induces 40x higher hashing costs (on the order of 25ms for a trie with only 1k leaves) and as a result they have been kept.
+
+An attempt to keep extension nodes for witness and not the merkelization rule can be found [here](https://notes.ethereum.org/m5VMkX8FRvi0Q_OOR7TF4A).
```

Comment:
When possible, try not to link externally.  Can the contents of this document be included inline or as an asset in `../assets/eip-3102/`?